sensor_msgs/include/sensor_msgs/fill_image.hpp

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2008, Willow Garage, Inc.
+// All rights reserved.
+//
+// Software License Agreement (BSD License 2.0)
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above
+//    copyright notice, this list of conditions and the following
+//    disclaimer in the documentation and/or other materials provided
+//    with the distribution.
+//  * Neither the name of the Willow Garage nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+// This file is originally ported from ROS1:
+// https://github.com/ros/common_msgs/blob/jade-devel/sensor_msgs/include/sensor_msgs/fill_image.h
+
+#ifndef SENSOR_MSGS__FILL_IMAGE_HPP_
+#define SENSOR_MSGS__FILL_IMAGE_HPP_
+
+#include <sensor_msgs/msg/image.hpp>
+#include <sensor_msgs/image_encodings.hpp>
+
+#include <string>
+
+namespace sensor_msgs
+{
+/**
+ * @brief Fills an image message.
+ * @param image Image to be filled.
+ * @param encoding_arg Encoding type, such as sensor_msgs::image_encodings::RGB8.
+ * @param rows_arg Number of rows.
+ * @param cols_arg Number of columns.
+ * @param step_arg Step size.
+ * @param data_arg Data to fill image with.
+ * @return True if successful.
```

Comment:
nitpick: Typically we use this style:

```c++
/// Fills an image message.
/**
 * \param image Image to be filled.
 * ...
 * \return True if successful.
 */
static inline bool fillImage(
  ...
```

Though we're not 💯percent consistent, here are some good examples (imo):

- https://github.com/ros2/rcutils/blob/f04cc666dd06347050ff5c18b80079922e729d92/include/rcutils/get_env.h#L26-L58
- https://github.com/ros2/rcl/blob/e5e338a8e0d0b4a4824edae7f351963ca2435015/rcl/include/rcl/validate_topic_name.h#L38-L88
- https://github.com/ros2/rclcpp/blob/b1af28047c26da0fe5723e3ea95ed1143bc224e5/rclcpp/include/rclcpp/node.hpp#L70-L76

I believe this is referred to as the "special C++ style comment" in the doxygen documentation (second instance of "option 3"):

https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html#cppblock

It avoids the `\brief`, and for documentation that only has a brief, it avoids `/** ... */` stuff.

Also we prefer the `\style` because that's what the doxygen documentation uses, e.g.:

https://www.stack.nl/~dimitri/doxygen/manual/commands.html

I think we somewhat mention this in our developer guide, but it could be clearer:

https://github.com/ros2/ros2/wiki/Developer-Guide#comments-and-doc-comments

Mostly just to be consistent with the rest of the newly written code.

Reply:
Yeah it's a matter of keeping the file close to its' ROS1 equivalent vs adopting the ROS2 style. It looks like the other utilities in sensor_msgs package (e.g. [point_cloud2_iterator.hpp](https://github.com/ros2/common_interfaces/blob/3d9f88ccdb65bb3a087db1066b17442e36b2a8cd/sensor_msgs/include/sensor_msgs/impl/point_cloud2_iterator.hpp)) kept the original documentation style (`@`), the conjugated version `Fills` instead of `Fill` etc

Reply:
The original file for this one doesn't have documentation at all, it was added in the migration.

The one you linked to mixes style anyways :/ https://github.com/ros2/common_interfaces/blob/3d9f88ccdb65bb3a087db1066b17442e36b2a8cd/sensor_msgs/include/sensor_msgs/impl/point_cloud2_iterator.hpp#L46

Reply:
You're right though it should really be `Fill` and not `Fills` (something we have tests for in Python but not C++), but since this is new documentation it's probably worth doing.

Reply:
> The original file for this one doesn't have documentation at all, it was added in the migration.

Yeah it's worth doing then, I just assumed it came from upstream